File: apps/AI-Keyboard/Android/app/src/main/java/ai/zetic/demo/cherrypad/model/Language.kt
package ai.zetic.demo.cherrypad.model

/**
 * A translation target language. [englishName] is used when building the model prompt;
 * [nativeName] is the endonym shown in the picker. The multilingual model infers the
 * source language from the text, so only the target is needed.
 */
data class Language(
    val id: String,          // stable identifier / rough BCP-47 code
    val englishName: String,
    val nativeName: String,
) {
    companion object {
        /** A broad set of languages the small model handles; ordered alphabetically. */
        val all: List<Language> = listOf(
            Language("ar", "Arabic", "العربية"),
            Language("bn", "Bengali", "বাংলা"),
            Language("zh-Hans", "Chinese (Simplified)", "简体中文"),
            Language("zh-Hant", "Chinese (Traditional)", "繁體中文"),
            Language("cs", "Czech", "Čeština"),
            Language("da", "Danish", "Dansk"),
            Language("nl", "Dutch", "Nederlands"),
            Language("en", "English (US)", "English (US)"),
            Language("fi", "Finnish", "Suomi"),
            Language("fr", "French", "Français"),
            Language("de", "German", "Deutsch"),
            Language("el", "Greek", "Ελληνικά"),
            Language("hi", "Hindi", "हिन्दी"),
            Language("id", "Indonesian", "Indonesia"),
            Language("it", "Italian", "Italiano"),
            Language("ja", "Japanese", "日本語"),
            Language("ko", "Korean", "한국어"),
            Language("ms", "Malay", "Melayu"),
            Language("nb", "Norwegian", "Norsk"),
            Language("fa", "Persian", "فارسی"),
            Language("pl", "Polish", "Polski"),
            Language("pt", "Portuguese", "Português"),
            Language("ro", "Romanian", "Română"),
            Language("ru", "Russian", "Русский"),
            Language("es", "Spanish", "Español"),
            Language("sv", "Swedish", "Svenska"),
            Language("th", "Thai", "ไทย"),
            Language("tr", "Turkish", "Türkçe"),
            Language("uk", "Ukrainian", "Українська"),
            Language("vi", "Vietnamese", "Tiếng Việt"),
        )

        /** Default translation target. */
        val default: Language get() = named("ko")

        fun named(id: String): Language =
            all.firstOrNull { it.id == id } ?: all.first { it.id == "en" }

        /** Lookup used when resolving a saved `targetLanguage` string. */
        fun matching(englishName: String): Language? =
            all.firstOrNull { it.englishName == englishName }
    }
}

File: apps/AI-Keyboard/iOS/CherryPad/Models/Language.swift
import Foundation

/// A translation target language. `englishName` is used when building the model
/// prompt; `nativeName` is the endonym shown in the picker. The multilingual Qwen
/// model infers the source language from the text, so only the target is needed.
struct Language: Identifiable, Hashable {
    let id: String          // stable identifier / rough BCP-47 code
    let englishName: String
    let nativeName: String
}

extension Language {
    /// A broad set of languages the small Qwen model handles; ordered alphabetically.
    static let all: [Language] = [
        Language(id: "ar", englishName: "Arabic", nativeName: "العربية"),
        Language(id: "bn", englishName: "Bengali", nativeName: "বাংলা"),
        Language(id: "zh-Hans", englishName: "Chinese (Simplified)", nativeName: "简体中文"),
        Language(id: "zh-Hant", englishName: "Chinese (Traditional)", nativeName: "繁體中文"),
        Language(id: "cs", englishName: "Czech", nativeName: "Čeština"),
        Language(id: "da", englishName: "Danish", nativeName: "Dansk"),
        Language(id: "nl", englishName: "Dutch", nativeName: "Nederlands"),
        Language(id: "en", englishName: "English (US)", nativeName: "English (US)"),
        Language(id: "fi", englishName: "Finnish", nativeName: "Suomi"),
        Language(id: "fr", englishName: "French", nativeName: "Français"),
        Language(id: "de", englishName: "German", nativeName: "Deutsch"),
        Language(id: "el", englishName: "Greek", nativeName: "Ελληνικά"),
        Language(id: "hi", englishName: "Hindi", nativeName: "हिन्दी"),
        Language(id: "id", englishName: "Indonesian", nativeName: "Indonesia"),
        Language(id: "it", englishName: "Italian", nativeName: "Italiano"),
        Language(id: "ja", englishName: "Japanese", nativeName: "日本語"),
        Language(id: "ko", englishName: "Korean", nativeName: "한국어"),
        Language(id: "ms", englishName: "Malay", nativeName: "Melayu"),
        Language(id: "nb", englishName: "Norwegian", nativeName: "Norsk"),
        Language(id: "fa", englishName: "Persian", nativeName: "فارسی"),
        Language(id: "pl", englishName: "Polish", nativeName: "Polski"),
        Language(id: "pt", englishName: "Portuguese", nativeName: "Português"),
        Language(id: "ro", englishName: "Romanian", nativeName: "Română"),
        Language(id: "ru", englishName: "Russian", nativeName: "Русский"),
        Language(id: "es", englishName: "Spanish", nativeName: "Español"),
        Language(id: "sv", englishName: "Swedish", nativeName: "Svenska"),
        Language(id: "th", englishName: "Thai", nativeName: "ไทย"),
        Language(id: "tr", englishName: "Turkish", nativeName: "Türkçe"),
        Language(id: "uk", englishName: "Ukrainian", nativeName: "Українська"),
        Language(id: "vi", englishName: "Vietnamese", nativeName: "Tiếng Việt"),
    ]

    static func named(_ id: String) -> Language {
        all.first { $0.id == id } ?? all.first { $0.id == "en" }!
    }

    /// Lookup used when resolving a handoff payload's `targetLanguage` string.
    static func matching(englishName: String) -> Language? {
        all.first { $0.englishName == englishName }
    }
}
